.buildkite/pipeline_generator/pipeline_generator.py

```diff
@@ -0,0 +1,180 @@
+        docker_image_path, 
+        test_bash_command, 
+        step.no_gpu
+    )
+
+def read_test_steps(file_path: str) -> List[TestStep]:
+    """Read test steps from test pipeline yaml and parse them into Step objects."""
+    with open(file_path, "r") as f:
+        content = yaml.safe_load(f)
+    return [TestStep(**step) for step in content["steps"]]
+
+def process_step(step: TestStep, run_all: str, list_file_diff: List[str]) -> List[Union[BuildkiteStep, BuildkiteBlockStep]]:
+    """Process test step and return corresponding BuildkiteStep."""
+    steps = []
+    current_step = BuildkiteStep(label=step.label, key=get_step_key(step.label), parallelism=step.parallelism, soft_fail=step.soft_fail, plugins=[_get_plugin_config(step)])
+    current_step.agents["queue"] = get_agent_queue(step.no_gpu, step.gpu, step.num_gpus).value
+
+    # If the step is a multi-node test, run it with a different script instead of Docker plugin
+    if step.num_nodes > 1:
+        current_step.commands = get_multi_node_test_command(
+            step.commands, 
+            step.working_dir, 
+            step.num_nodes, 
+            step.num_gpus, 
+            get_image_path()
+        )
+        current_step.plugins = None
+
+    def step_should_run():
+        """Determine whether the step should automatically run or not."""
+        if not step.source_file_dependencies or run_all == "1" or step.gpu == "a100":
+            return True
+        return any(source_file in diff_file 
+            for source_file in step.source_file_dependencies 
+            for diff_file in list_file_diff)
+
+    # If step does not have to run automatically, create a block step
+    # and have the current step depend on it
+    if not step_should_run():
+        block_step = get_block_step(step.label)
+        steps.append(block_step)
+        current_step.depends_on = block_step.key
+    steps.append(current_step)
+    return steps
+
+def generate_build_step() -> BuildkiteStep:
+    """Build the Docker image and push it to ECR."""
+    buildkite_commit = os.getenv("BUILDKITE_COMMIT")
+    docker_image = get_image_path()
+    build_commands = [
+        "aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws/q9t5s3a7",
+        f"docker build --build-arg max_jobs=16 --build-arg buildkite_commit={buildkite_commit} --build-arg USE_SCCACHE=1 --tag {docker_image} --target test --progress plain .", 
+        f"docker push {docker_image}"
+    ]
+    step = BuildkiteStep(
+        label=":docker: build image", 
+        key="build", 
+        agents={"queue": AgentQueue.AWS_CPU.value}, 
+        env={"DOCKER_BUILDKIT": "1"}, 
+        retry={"automatic": [{"exit_status": -1, "limit": 2}, {"exit_status": -10, "limit": 2}]}, 
+        commands=build_commands,
+        depends_on=None,
+    )
+    return step
+
+def get_external_hardware_tests(test_steps: List[TestStep]) -> List[Union[BuildkiteStep, BuildkiteBlockStep]]:
```

Comment:
feels that this is very specific to AMD? like `DOCKER_IMAGE_AMD` is hard-coded now?

Reply:
yes cause the image path changes by commit (it's tagged with commit)